pygeo/parameterization/DVGeoCST.py

```diff
@@ -0,0 +1,1175 @@
+        if distance < distTol:
+            self.foilCoords = self.foilCoords[:-1, :]
+
+        # Traverse the airfoil surface to find the corner(s) defining the trailing edge (ignore anything in the front
+        # half, chordwise, of the airfoil)
+        cornerIdx = []
+        for idx in range(self.foilCoords.shape[0]):
+            pt = self.foilCoords[idx, :]
+            # Ignore if closer to the leading edge
+            if pt[self.xIdx] - self.xMin < self.xMax - pt[self.xIdx]:
+                continue
+            edgePrev = pt - self.foilCoords[idx - 1, :]
+            edgePrev /= np.linalg.norm(edgePrev)
+            edgeNext = self.foilCoords[(idx + 1) % self.foilCoords.shape[0], :] - pt
+            edgeNext /= np.linalg.norm(edgeNext)
+            if np.dot(edgePrev, edgeNext) < tolTE:
+                cornerIdx.append(idx)
+        if len(cornerIdx) > 2:
+            raise RuntimeError(
+                "More than two corners in the airfoil identified when looking for the "
+                + "trailing edge. If the actual airfoil geometry in the dat file has more "
+                + "than two corners it is not supported. If not, try reducing the tolTE input."
+            )
+        elif len(cornerIdx) == 0:
+            raise RuntimeError(
+                "Zero corners in the airfoil identified when looking for the "
+                + "trailing edge. If the actual airfoil geometry in the dat file has zero corners "
+                + "(is a circle??) it is not supported. If not, try increasing the tolTE input."
+            )
+
+        # Airfoil is sharp if only one corner is detected
+        self.sharp = len(cornerIdx) == 1
+
+        # Save the upper and lower trailing edge coordinates if it is not sharp
+        if self.sharp:
+            self.thicknessTE = np.array([0.0])
+        else:
+            if self.foilCoords[cornerIdx[0], self.yIdx] > self.foilCoords[cornerIdx[1], self.yIdx]:
+                self.coordUpperTE = self.foilCoords[cornerIdx[0]]
+                self.coordLowerTE = self.foilCoords[cornerIdx[1]]
+            else:
+                self.coordUpperTE = self.foilCoords[cornerIdx[1]]
+                self.coordLowerTE = self.foilCoords[cornerIdx[0]]
+            self.thicknessTE = self.coordUpperTE[self.yIdx] - self.coordLowerTE[self.yIdx]
+
+        # Compute splines for the upper and lower surfaces (used to split the foil in addPointSet).
+        # preFoil defines the leading edge as the point furthest from the trailing edge
+        self.foil = Airfoil(coords)
+        self.upperSpline, self.lowerSpline = self.foil.splitAirfoil()
+
+        # Fit CST parameters to the airfoil's upper and lower surface
+        self.idxFoil = {}
+        self.idxFoil["upper"], self.idxFoil["lower"] = self._splitUpperLower(self.foilCoords)
+        chord = self.xMax - self.xMin
+        self.defaultDV["chord"][0] = chord
+        print(f"######## Fitting CST coefficients to coordinates in {datFile} ########")
```

Comment:
is this printed on all procs? Also related; is the fit done in parallel on all procs or do you do it on once and communicate the result? if you are doing it in parallel, there's a small chance the procs will get different values, especially when the fit is ill conditioned